File: Python/Interleaving String.py
class Solution:
    # @return a boolean
    def isInterleave(self, s1, s2, s3):
        if len(s1)+len(s2) != len(s3): 
            return False
        dp = [[False for j in range(len(s2)+1)] for i in range(len(s1)+1)]

        for i in range(0, len(s1)+1):
            for j in range(0, len(s2)+1):
                if i == 0 and j == 0:
                    dp[i][j] = True
                elif i == 0:                        
                    dp[i][j] = s2[:j] == s3[:j]
                elif j == 0:                        
                    dp[i][j] = s1[:i] == s3[:i]
                else:
                    dp[i][j] = (dp[i-1][j] and s1[i-1] == s3[i+j-1]) or (dp[i][j-1] and s2[j-1] == s3[i+j-1])
        return dp[-1][-1]

File: Java/src/Interleaving String.java
public class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
         if ((s1.length()+s2.length())!=s3.length()) return false;

        boolean[][] dp = new boolean[s2.length()+1][s1.length()+1];
    
        for (int i = 0; i < dp.length; i++){
            for (int j = 0; j < dp[0].length; j++){
            	if (i== 0 && j==0) {
					dp[i][j] = true;
				}
            	else if (i==0) {
            		dp[i][j] = dp[i][j-1]&&(s1.charAt(j-1)==s3.charAt(j-1));
				}
            	else if (j==0) {
            		 dp[i][j] = dp[i-1][j]&&(s2.charAt(i-1)==s3.charAt(i-1));
				}
            	else {
                    dp[i][j] = (dp[i-1][j]&&(s2.charAt(i-1)==s3.charAt(i+j-1)))
                    || (dp[i][j-1]&&(s1.charAt(j-1)==s3.charAt(i+j-1)));
				}
            }
        }    
        return dp[s2.length()][s1.length()];
    }
}